xfrd.c

```diff
@@ -1191,86 +1206,63 @@ xfrd_set_timer(xfrd_zone_type* zone, time_t t)
 }
 
 void
-xfrd_handle_incoming_soa(xfrd_zone_type* zone,
-	xfrd_soa_type* soa, time_t acquired)
+xfrd_handle_incoming_soa(
+	xfrd_zone_type* zone,
+	xfrd_soa_type* soa,
+	uint32_t serial,
+	time_t acquired,
+	xfrd_xfr_state_type state)
 {
-	if(soa == NULL) {
-		/* nsd no longer has a zone in memory */
-		zone->soa_nsd_acquired = 0;
-		xfrd_set_zone_state(zone, xfrd_zone_refreshing);
-		xfrd_set_refresh_now(zone);
-		return;
-	}
-	if(zone->soa_nsd_acquired && soa->serial == zone->soa_nsd.serial)
-		return;
+	xfrd_xfr_type *xfr = zone->latest_xfr;
 
-	if(zone->soa_disk_acquired && soa->serial == zone->soa_disk.serial)
-	{
-		/* soa in disk has been loaded in memory */
-		log_msg(LOG_INFO, "zone %s serial %u is updated to %u",
-			zone->apex_str, (unsigned)ntohl(zone->soa_nsd.serial),
-			(unsigned)ntohl(soa->serial));
-		zone->soa_nsd = zone->soa_disk;
-		zone->soa_nsd_acquired = zone->soa_disk_acquired;
-		xfrd->write_zonefile_needed = 1;
-		/* reset exponential backoff, we got a normal timer now */
-		zone->fresh_xfr_timeout = 0;
-		if(xfrd_time() - zone->soa_disk_acquired
-			< (time_t)ntohl(zone->soa_disk.refresh))
-		{
-			/* zone ok, wait for refresh time */
-			xfrd_set_zone_state(zone, xfrd_zone_ok);
-			zone->round_num = -1;
-			xfrd_set_timer_refresh(zone);
-		} else if(xfrd_time() - zone->soa_disk_acquired
-			< (time_t)ntohl(zone->soa_disk.expire))
-		{
-			/* zone refreshing */
-			xfrd_set_zone_state(zone, xfrd_zone_refreshing);
-			xfrd_set_refresh_now(zone);
-		}
-		if(xfrd_time() - zone->soa_disk_acquired
-			>= (time_t)ntohl(zone->soa_disk.expire)) {
-			/* zone expired */
-			xfrd_set_zone_state(zone, xfrd_zone_expired);
-			xfrd_set_refresh_now(zone);
+	if(soa == NULL) {
+		/* soa information is missing if a zone transfer is corrupt, or
+		   a zone was read from disk. if a zone transfer corrupts the
+		   database the new server "aborts", in which case the difffile
+		   is marked discarded and is skipped next reload */
+		if(xfr == NULL) {
+			assert(serial == 0);
+			assert(state == xfrd_xfr_new);
+			zone->soa_nsd_acquired = 0;
+			memset(&zone->soa_nsd, 0, sizeof(xfrd_soa_type));
+			return;
 		}
 
-		if(zone->soa_notified_acquired != 0 &&
-			(zone->soa_notified.serial == 0 ||
-		   	compare_serial(ntohl(zone->soa_disk.serial),
-				ntohl(zone->soa_notified.serial)) >= 0))
-		{	/* read was in response to this notification */
-			zone->soa_notified_acquired = 0;
-		}
-		if(zone->soa_notified_acquired && zone->state == xfrd_zone_ok)
+		assert(serial != 0);
```

Comment:
This could be 0 right? In an actual zone transfer, 0 serial is possible in a SOA record.